algorithms/kernel/svm/svm_train_thunder_impl.i

```diff
@@ -307,29 +342,64 @@ services::Status SVMTrainImpl<thunder, algorithmFPType, ParameterType, cpu>::SMO
 {
-    char trans = 'N';
-    DAAL_INT m = nVectors;
-    DAAL_INT n = nWS;
-    algorithmFPType alpha(1.0);
-    DAAL_INT ldA = m;
+    SafeStatus safeStat;
+    const size_t nBlocksWS = nWS / _blockSizeWS;
+
+    size_t blockSize   = 256;
+    size_t nBlocksGrad = (nVectors / blockSize) + !!(nVectors % blockSize);
+
     DAAL_INT incX(1);
-    algorithmFPType beta(1.0);
     DAAL_INT incY(1);
 
-    Blas<algorithmFPType, cpu>::xgemv(&trans, &m, &n, &alpha, kernelWS, &ldA, deltaalpha, &incX, &beta, grad, &incY);
+    daal::threader_for(nBlocksWS, nBlocksWS, [&](const size_t iBlock) {
+        const size_t startRowWS           = iBlock * _blockSizeWS;
+        const algorithmFPType deltaalphai = deltaalpha[startRowWS];
+
+        daal::threader_for(nBlocksGrad, nBlocksGrad, [&](const size_t iBlockGrad) {
+            const size_t nRowsInBlockGrad = (iBlockGrad != nBlocksGrad - 1) ? blockSize : nVectors - iBlockGrad * blockSize;
+            const size_t startRowGrad     = iBlockGrad * blockSize;
+            algorithmFPType * gradi       = &gradBuff[nVectors * iBlock + startRowGrad];
+
+            ReadColumns<algorithmFPType, cpu> mtKernel(kernelWS.get(), startRowWS, startRowGrad, nRowsInBlockGrad);
+            DAAL_CHECK_BLOCK_STATUS_THR(mtKernel);
+            const algorithmFPType * kernelBlockI = mtKernel.get();
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t j = 0; j < nRowsInBlockGrad; j++)
+            {
+                gradi[j] = deltaalphai * kernelBlockI[j];
+            }
+
+            for (size_t i = 1; i < _blockSizeWS; i++)
+            {
+                ReadColumns<algorithmFPType, cpu> mtKernel(kernelWS.get(), startRowWS + i, startRowGrad, nRowsInBlockGrad);
+                DAAL_CHECK_BLOCK_STATUS_THR(mtKernel);
+                algorithmFPType * kernelBlockI = (algorithmFPType *)mtKernel.get();
+                algorithmFPType deltaalphai    = deltaalpha[startRowWS + i];
+                Blas<algorithmFPType, cpu>::xxaxpy((DAAL_INT *)&nRowsInBlockGrad, &deltaalphai, kernelBlockI, &incX, gradi, &incY);
+            }
+        });
+    });
+
+    algorithmFPType one = algorithmFPType(1);
+    for (size_t i = 0; i < nBlocksWS; i++)
+    {
+        Blas<algorithmFPType, cpu>::xxaxpy((DAAL_INT *)&nVectors, &one, &gradBuff[i * nVectors], &incX, grad, &incY);
+    }
+
     return services::Status();
 }
 
 template <typename algorithmFPType, typename ParameterType, CpuType cpu>
 bool SVMTrainImpl<thunder, algorithmFPType, ParameterType, cpu>::checkStopCondition(const algorithmFPType diff, const algorithmFPType diffPrev,
                                                                                     const algorithmFPType eps, size_t & sameLocalDiff)
 {
-    sameLocalDiff = internal::Math<algorithmFPType, cpu>::sFabs(diff - diffPrev) < eps * 1e-2 ? sameLocalDiff + 1 : 0;
-
+    sameLocalDiff = internal::Math<algorithmFPType, cpu>::sFabs(diff - diffPrev) < eps * 1e-3 ? sameLocalDiff + 1 : 0;
```

Comment:
where did this constant come from?

Reply:
We have two critera for stop. 1. `diff < eps`. 2. diff stays unchanged. `nNoChanges` - number of repetitions.  It is not fair to compare repetitions with the same parameter `eps`. It should rather be a very small number. Moreover, this criterion will rarely work in practice - in situations when the gradient, for example, is unstable calculated. This number - 1e-3 is taken from observations.